### PR TITLE
Feature/ch805/elasticsearch as search backend

### PR DESCRIPTION
**Pull request by DarkSnow**
Replaces the database directory search with an Elastic Search equivalent.

WARNING: This will remove the DB based view from your config.

This also encompasses some of the changes to the operation of the directory discussed in the new design work.

### CODE REVIEW

**Comment by joesb** [comment]
I was about to start looking at this, but there's no review requested and I can't see a ticket in 'Ready for Review' to work from. This branch is named for [ch805], but that's not in this sprint. [ch951] is still in doing. Is it 951 that's been the 'container' for this work?

**Comment by DarkSnow** [comment]
Not sure what's happening there Joe. I'm importing the config fine and the two settings for facet summary as saying facets_summary will not  be installed, but it's in core.extension.yml so it will be activated and it's parent module, facets, is in composer.json

**Comment by joesb** [comment]
Well, weird …

```sh
make down
make up
drush cim
```

all importing fine.

**Comment by joesb** [comment]
Above problem solved by selecting an updated elasticsearch, e.g. `ELASTICSEARCH_TAG=7.4-5.2.3 `

**Comment by joesb** [comment]
New problem:

> The website encountered an unexpected error. Please try again later.
> 
> _Symfony\Component\Routing\Exception\RouteNotFoundException_: Route "view.directory.page_1" does not exist.

Just needs an update in https://github.com/cabinetoffice/national-leadership-centre-drupal/blob/feature/ch805/elasticsearch-as-search-backend/web/modules/custom/nlc_prototype/src/Controller/DirectoryTokenAccessConfirmController.php#L33

**Comment by joesb** [comment]
So it seems that this branch is Elasticsearch _and_ lots of the UI changes. I wonder if it's possible to extract the two parts into separate branches … ?

1) Elasticsearch addition and rebuild of the current View and facets to work with the ES search index backend
2) a branch based on 1) (rather than master) that has the UI changes and more in it.

That way we can, potentially, merge and deploy 1) without being blocked by work on 2).

Does that make sense? Is that possible, to create branch 1) from a particular commit hash at some point up the stack? And then this branch can become branch 2).

File: web/modules/custom/nlc_prototype/src/Plugin/Block/MyCohortBlock.php [hunk 1]
```diff
@@ -42,8 +42,8 @@ public function build() {
           $entity = \Drupal::entityTypeManager()
             ->getStorage('node')
             ->load($cohort->target_id);
-          $options['query']['directory'][] = 'cohort:' . $cohort->target_id;
-          if (!empty($options)) {
+          $options['query']['f'][] = 'cohort:' . $entity->label();
+          if ($entity && !empty($options)) {
```
Comment (joesb) [inline web/modules/custom/nlc_prototype/src/Plugin/Block/MyCohortBlock.php]: Basically, this all looks great, but just one tiny question here. The `if` test includes the entity object, but in the line above the `$entity->label()` method has just been used — so this tests for the existence of something that's already been  assumed to exist. Which is a bit odd.

Worth tidying up?

File: .env.example [hunk 2]
```diff
@@ -42,6 +42,10 @@ SALESFORCE_URL=https://salesforce.com
 # Salesforce login user
 SALESFORCE_LOGIN=user@nationalleadership.gov.uk
 
+# The URL of the Elastic Search Cluster
+ELASTICSEARCH_URL=http://nlc_cms_elasticsearch:9200
```
Comment (joesb) [inline .env.example]: I can see this env var being created here in this commit, but no use of that in a file, like e.g. `settings.php` — should something else be committed also?

**Comment by joesb** [comment]
Now #95 is merged, this branch/PR has so conflicts so will need to have `master` merged in.

**Comment by joesb** [comment]
@DarkSnow this branch now has some merge conflicts with `master`.

Also, `master` now includes the search results summary block work you did — does the follow-up work you did on the prettified Elasticsearch work now supersede this branch?

Could you find some time to take a look at the merge conflicts and/or adding a PR to the follow-on Elasticsearch work? When we get the ITHC back, we'll be looking at how quickly we can make any necessary fixes and move to AWS so we can get into 'public beta' and add all the extra Network individuals before the NLF event.

**Comment by DarkSnow** [comment]
I think since this has been delayed the code will quickly go out of date.

The result summary work was independent of this, related to the facet_summary module, so I pulled it all out and created a new summary block for the DB search. Similarly, the collapsible block work relates to the blocks but not the search as such. These conflicts have come from work that overlaps this module.

When the time comes to release we'll need to check that the correct blocks (summary and facets) are in the correct place, that the DB blocks have been removed and that any collapsiblock settings are in place for these new blocks.

So, there will be work to do to get this ready when we get the go ahead.

I've merged master back in again, but further testing should be delayed until we're ready to deploy, just to save repeating the tests and wasting time.

**Comment by joesb** [comment]
Sadly, merge of #102 gives a merge conflict to resolve here.

**Comment by joesb** [comment]
Right, finally merging this work to `master` – HOORAY!

**Pull request merged**